components/net/http_cache.rs

```diff
@@ -0,0 +1,486 @@
+                    let has_public_directive = directives.iter().any(|directive| {
+                        header::CacheDirective::Public == *directive
+                    });
+                    if has_public_directive {
+                        return heuristic_freshness;
+                    }
+                }
+            },
+        }
+    }
+    // Requires validation upon first use as default.
+    Duration::seconds(0i64)
+}
+
+/// Request Cache-Control Directives
+/// <https://tools.ietf.org/html/rfc7234#section-5.2.1>
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    let directive_data = match request.headers.get_raw("cache-control") {
+        Some(data) => data,
+        None => return expires,
+    };
+    let directives_string = String::from_utf8_lossy(&directive_data[0]);
+    for directive in directives_string.split(",") {
+        let mut directive_info = directive.split("=");
+        match (directive_info.nth(0), directive_info.nth(0)) {
+            (Some("max-stale"), Some(sec_str)) => {
+                if let Ok(secs) = sec_str.parse::<i64>() {
+                    return expires + Duration::seconds(secs);
+                }
+            },
+            (Some("max-age"), Some(sec_str)) => {
+                if let Ok(secs) = sec_str.parse::<i64>() {
+                    let max_age = Duration::seconds(secs);
+                    if expires > max_age {
+                        return Duration::min_value();
+                    }
+                    return expires - max_age;
+                }
+            },
+            (Some("min-fresh"), Some(sec_str)) => {
+                if let Ok(secs) = sec_str.parse::<i64>() {
+                    let min_fresh = Duration::seconds(secs);
+                    if expires < min_fresh {
+                        return Duration::min_value();
+                    }
+                    return expires - min_fresh;
+                }
+            },
+            (Some("no-cache"), _) | (Some("no-store"), _) => return Duration::min_value(),
+            _ => {}
+        }
+    }
+    expires
+}
+
+/// Create a CachedResponse from a request and a CachedResource.
+fn create_cached_response(request: &Request, cached_resource: &CachedResource, cached_headers: &Headers)
+    -> CachedResponse {
+    let mut response = Response::new(cached_resource.metadata.final_url.clone());
+    response.headers = cached_headers.clone();
+    response.body = cached_resource.body.clone();
+    response.status = cached_resource.status.clone();
+    response.raw_status = cached_resource.raw_status.clone();
+    response.url_list = cached_resource.url_list.clone();
+    response.https_state = cached_resource.https_state.clone();
+    response.referrer = cached_resource.referrer.clone();
+    response.referrer_policy = cached_resource.referrer_policy.clone();
```

Comment:
This doesn't look right - the referrer and referrer policy can vary in the request that's being made; we shouldn't be caching them and reusing them later.